pubsub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/GooglePubSubSink.scala

```diff
@@ -100,34 +111,49 @@ object GooglePubSubSink {
     for {
       batching <- batchingSettings(bufferConfig).asRight
       retry = retrySettings(googlePubSubConfig.backoffPolicy)
-      publisher <- createPublisher(googlePubSubConfig.googleProjectId, topicName, batching, retry)
-      _ <- if (enableStartupChecks) topicExists(googlePubSubConfig.googleProjectId, topicName).flatMap { b =>
-        if (b) ().asRight
-        else new IllegalArgumentException(s"Google PubSub topic $topicName doesn't exist").asLeft
+      customProviders = sys.env.get("PUBSUB_EMULATOR_HOST").map { hostPort =>
+        val channel             = ManagedChannelBuilder.forTarget(hostPort).usePlaintext().build()
+        val channelProvider     = FixedTransportChannelProvider.create(GrpcTransportChannel.create(channel))
+        val credentialsProvider = NoCredentialsProvider.create()
+        (channelProvider, credentialsProvider)
+      }
+      _ <- if (enableStartupChecks) {
+        val topicAdmin = createTopicAdmin(customProviders)
+        try {
+          topicExists(topicAdmin, googlePubSubConfig.googleProjectId, topicName)
+        } finally {
+          Either.catchNonFatal(topicAdmin.close())
+          ()
+        }
       } else ().asRight
+      publisher <- createPublisher(googlePubSubConfig.googleProjectId, topicName, batching, retry, customProviders)
     } yield new GooglePubSubSink(maxBytes, publisher, topicName)
 
   private val UserAgent = s"snowplow/stream-collector-${generated.BuildInfo.version}"
 
   /**
-    * Instantiates a Publisher on an existing topic with the given configuration options.
+    * Instantiates a Publisher on a topic with the given configuration options.
     * This can fail if the publisher can't be created.
     * @return a PubSub publisher or an error
     */
   private def createPublisher(
     projectId: String,
     topicName: String,
     batchingSettings: BatchingSettings,
-    retrySettings: RetrySettings
-  ): Either[Throwable, Publisher] =
-    Either.catchNonFatal(
-      Publisher
-        .newBuilder(ProjectTopicName.of(projectId, topicName))
-        .setBatchingSettings(batchingSettings)
-        .setRetrySettings(retrySettings)
-        .setHeaderProvider(FixedHeaderProvider.create("User-Agent", UserAgent))
-        .build()
-    )
+    retrySettings: RetrySettings,
+    customProviders: Option[(TransportChannelProvider, CredentialsProvider)]
+  ): Either[Throwable, Publisher] = {
+    val builder = Publisher
+      .newBuilder(ProjectTopicName.of(projectId, topicName))
+      .setBatchingSettings(batchingSettings)
+      .setRetrySettings(retrySettings)
+      .setHeaderProvider(FixedHeaderProvider.create("User-Agent", UserAgent))
+    val customized = customProviders.fold(builder) {
```

Comment:
We do this kind of thing a lot in snowplow code, where we take the java builder pattern and then we pretend that the builder is an immutable val.  We force ourselves to use things like `.fold` so we can say we're being functional.   I wonder though if it's better to just be honest that builders are mutable.  And instead do something like this:

```scala
val builder = Publisher
      .newBuilder(ProjectTopicName.of(projectId, topicName))
  // lines skipped for brevity

customProviders.foreach { credentialsProvider =>
  case (channelProvider, credentialsProvider) =>
        builder.setChannelProvider(channelProvider).setCredentialsProvider(credentialsProvider)
}

builder.build()
```

Reply:
> we pretend that the builder is an immutable val

I actually forgot that it is not the case, good point!